### PR TITLE
dyno: implement opaque extern types

### PR DESCRIPTION
**Pull request by DanilaFe**
This PR implements opaque extern types in Dyno. An opaque extern type is one defined as follows:

```Chapel
extern type myType;
```

Types like these are "opaque" in the sense that Chapel doesn't know their underlying representation; the name is pretty much the only identifying information about such types.

This PR diverges from production compiler behavior in that extern types with the same name are considered equal, even if they were declared in different places. In other words,

```Chapel
proc f1() type {
  extern type myType;
  return myType;
}

proc f2() type {
  extern type myType;
  return myType;
}

param x = f1() == f2(); // false in production, true with this PR
```

Reviewed by @benharsh -- thanks!

## Testing
- [x] paratest

### CODE REVIEW

File: frontend/include/chpl/types/ExternType.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+#include "chpl/types/Type.h"
+#include "chpl/types/QualifiedType.h"
+
+namespace chpl {
+namespace types {
+
+class ExternType final : public Type {
+ private:
+  UniqueString linkageName_;
```
[the remark this answers is not in the document]
Comment (benharsh) [reply frontend/include/chpl/types/ExternType.h:31]: Adding a test that checks the stringify'd name of extern types would also be useful.

**Review by benharsh — COMMENTED** [review]
Generally looks good, but I think we need to address the linkage-name vs. original-name question.

File: frontend/include/chpl/types/ExternType.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (DanilaFe) [reply frontend/include/chpl/types/ExternType.h:31]: Discussed this out of band, and Ben and I agreed that we can merge the PR as-is for the time being, and figure out how to keep the original name as part of the type later. The reason this doesn't immediately work is that our type comparisons in `canPass` are based on _pointer_ equality.

[the remark this answers is not in the document]
Comment (DanilaFe) [reply frontend/include/chpl/types/ExternType.h:31]: This is also tracked here: https://github.com/chapel-lang/chapel/issues/22711

**Pull request merged**